tests/test_sdeint.py

```diff
@@ -35,152 +36,149 @@
+    # Tests that sdeint:
+    # (a) runs/raises an error as appropriate
+    # (b) produces tensors of the right shape
+    # (c) accepts every method
+
+    if logqp:
+        pytest.skip("Temporarily deprecating logqp.")
+
+    should_fail = False
+    if sde_type == 'ito':
+        if method in ('euler_heun', 'heun', 'midpoint', 'log_ode'):
+            should_fail = True
+    else:
+        if method in ('euler', 'srk'):
+            should_fail = True
+    if method in ('milstein', 'srk') and sde_cls.noise_type == 'general':
+        should_fail = True
+    if method == 'srk' and levy_area_approximation == 'none':
+        should_fail = True
+    if method == 'log_ode' and levy_area_approximation in ('none', 'space-time'):
+        should_fail = True
+    if method == 'blah':
+        should_fail = True
+
+    if sde_cls is basic_sde.ScalarSDE:
+        kwargs = {'d': d}
+    elif sde_cls is basic_sde.DiagonalSDE:
+        kwargs = {}
+    else:
+        kwargs = {'d': d, 'm': m}
+    sde = sde_cls(sde_type=sde_type, **kwargs).to(device)
+
+
+    if use_bm:
+        if sde_cls.noise_type == 'scalar':
+            size = (batch_size, 1)
+        elif sde_cls.noise_type == 'diagonal':
+            size = (batch_size, d)
+        else:
+            assert sde_cls.noise_type in ('additive', 'general')
+            size = (batch_size, m)
+        bm = torchsde.BrownianInterval(t0=t0, t1=t1, size=size, dtype=dtype, device=device,
+                                       levy_area_approximation=levy_area_approximation)
+    else:
+        bm = None
+
+    _test_sdeint(sde, bm, method, adaptive, logqp, device, should_fail)
+
+
+@pytest.mark.parametrize("sde_cls", [basic_sde.BasicSDE1, basic_sde.BasicSDE2, basic_sde.BasicSDE3,
+                                     basic_sde.BasicSDE4])
+@pytest.mark.parametrize('method', ['euler', 'milstein', 'srk'])
+@pytest.mark.parametrize('adaptive', [False, True])
+@pytest.mark.parametrize('device', devices)
+def test_sdeint_dependencies(sde_cls, method, adaptive, device):
+    # This test uses diagonal noise. This checks if the solvers still work when some of the functions don't depend on
+    # the states/params and when some states/params don't require gradients.
+    sde = sde_cls(d=d).to(device)
+    bm = None
+    logqp = False
+    should_fail = False
+    _test_sdeint(sde, bm, method, adaptive, logqp, device, should_fail)
+
+
+def _test_sdeint(sde, bm, method, adaptive, logqp, device, should_fail):
+    y0 = torch.ones(batch_size, d, device=device)
+    ts = torch.linspace(t0, t1, steps=T, device=device)
+    if adaptive and method == 'euler' and sde.noise_type != 'additive':
+        ctx = pytest.warns(UserWarning)
+    else:
+        ctx = _nullcontext()
+
+    # Using `f` as drift.
+    with torch.no_grad():
+        try:
+            with ctx:
+                ans = torchsde.sdeint(sde, y0, ts, bm, method=method, dt=dt, adaptive=adaptive)
+        except Exception:
```

Comment:
There's something subtle here which is the exception raised might not be the exception we expect. I think catching an exception with a more specific type will help. Most cases that we raise explicitly are of ValueError type.

Reply:
Yup, you're right. I've now also tweaked a couple asserts to now raise ValueErrors instead.
(Shouldn't be a backward compatibility issue, I think the asserts aren't in master.)